src/ripple/app/tx/impl/apply.cpp

```diff
@@ -86,15 +93,79 @@ invoke_apply (ApplyContext& ctx)
 
 //------------------------------------------------------------------------------
 
+std::pair<Validity, std::string>
+checkValidity(HashRouter& router, STTx const& tx, bool allowMultiSign)
+{
+    auto const id = tx.getTransactionID();
+    auto const flags = router.getFlags(id);
+    if (flags & SF_SIGBAD)
+        // Signature is known bad
+        return std::make_pair(Validity::SigBad,
+            "Transaction has bad signature.");
+    if (!(flags & SF_SIGGOOD))
+    {
+        // Don't know signature state. Check it.
+        if (!tx.checkSign(allowMultiSign))
```

Comment:
Should `checkSign` really be a member of `STTx`? That means it would have to call back into `src/ripple/app` to handle SusPays or include `apply.tx`, breaking levelization (`src/ripple/protocol` does not include anything in `src/ripple/app`).